packages/labs/analyzer/src/lib/javascript/mixins.ts

```diff
@@ -0,0 +1,194 @@
+ *
+ * If the function is unannotated and does not match the above mixin shape, it
+ * will silently just be analyzed as a simple function and not a mixin. However,
+ * the `@mixin` annotation can be added to produce specific diagnostic errors
+ * when a condition for being analyzed as a mixin is not met.
+ */
+export const maybeGetMixinFromFunctionLike = (
+  fn: ts.FunctionLikeDeclaration,
+  name: ts.Identifier,
+  analyzer: Analyzer,
+  hasMixinHint = nodeHasMixinHint(fn)
+) => {
+  if (!fn.parameters || fn.parameters.length < 1) {
+    return throwIfMixin(
+      hasMixinHint,
+      fn,
+      `Expected mixin to have a superClass parameter.`
+    );
+  }
+  const possibleSuperClasses = fn.parameters.map((p) =>
+    ts.isIdentifier(p.name) ? p.name.text : ''
+  );
+  const functionBody = fn.body;
+  if (functionBody === undefined) {
+    return throwIfMixin(
+      hasMixinHint,
+      fn,
+      `Expected mixin to have a block function body.`
+    );
+  }
+  if (!ts.isBlock(functionBody)) {
+    return throwIfMixin(
+      hasMixinHint,
+      functionBody,
+      `Expected mixin to have a block function body; arrow-function class ` +
+        `expression syntax is not supported.`
+    );
+  }
+  let classDeclaration!: ts.ClassDeclaration;
+  let returnStatement!: ts.ReturnStatement;
+  functionBody.statements.forEach((s) => {
+    if (ts.isClassDeclaration(s)) {
+      classDeclaration = s;
+    }
+    if (ts.isReturnStatement(s)) {
+      returnStatement = s;
+    }
+  });
+  if (!classDeclaration) {
+    return throwIfMixin(
+      hasMixinHint,
+      functionBody,
+      `Expected mixin to contain a class declaration statement.`
+    );
+  }
+  if (!returnStatement) {
+    return throwIfMixin(
+      hasMixinHint,
+      functionBody,
+      `Expected mixin to contain a return statement returning a class.`
+    );
+  }
+  const extendsClause = classDeclaration.heritageClauses?.find(
+    (c) => c.token === ts.SyntaxKind.ExtendsKeyword
+  );
+  if (extendsClause === undefined) {
+    return throwIfMixin(
+      hasMixinHint,
+      classDeclaration,
+      `Expected mixin to contain class declaration extending a superClass argument to function.`
+    );
+  }
+  if (extendsClause.types.length !== 1) {
+    throw new DiagnosticsError(
+      extendsClause,
+      'Internal error: did not expect a mixin class extends clause to have multiple types'
+    );
+  }
+  const superClassArgIdx = findSuperClassArgIndexFromHeritage(
+    possibleSuperClasses,
+    extendsClause.types[0].expression
+  );
+  if (superClassArgIdx < 0) {
+    throw new DiagnosticsError(
+      extendsClause,
+      'Did not find a "superClass" argument used in the extends clause of mixin class.'
+    );
+  }
+
+  return new MixinDeclaration({
+    node: fn,
+    name: name.getText(),
+    superClassArgIdx,
+    classDeclaration: getClassDeclaration(classDeclaration, analyzer),
+  });
+};
+
+const findSuperClassArgIndexFromHeritage = (
```

Comment:
Comments on what's support here would be great too. I'm wondering about cases like:

```ts
const M = (A, B) => class extends A(B) { ... }
```

Seems like that will return `1` because `B` is the argument? I guess that's about right, though we do you return `A` as another applied mixin anywhere? I think it's fine if this is case we don't support, btw. It'd be nice if we could say that this first has to be inferred as a mixin by typescript.